src/components/message.js

```diff
@@ -4,159 +4,144 @@
+import { formatMessage as baseFormatMessage } from '../format'
 
 const defaultFormatMessage = (descriptor, values) => {
   if (process.env.NODE_ENV !== 'production') {
     console.error(
       '[React Intl] Could not find required `intl` object. <IntlProvider> needs to exist in the component ancestry. Using default message as fallback.'
-    );
+    )
   }
 
   return baseFormatMessage(
     {},
-    {getMessageFormat: memoizeIntlConstructor(IntlMessageFormat)},
+    { getMessageFormat: memoizeIntlConstructor(IntlMessageFormat) },
     descriptor,
     values
-  );
-};
-
-class FormattedMessage extends Component {
-  static displayName = 'FormattedMessage';
-
-  static propTypes = {
-    ...messageDescriptorPropTypes,
-    intl: intlShape,
-    values: PropTypes.object,
-    tagName: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
-    children: PropTypes.func,
-  };
-
-  static defaultProps = {
-    values: {},
-  };
-
-  constructor(props) {
-    super(props);
-    if (!props.defaultMessage) {
-      invariantIntlContext(props);
-    }
+  )
+}
+
+export function arePropsEqual(props, nextProps) {
+  const { values } = props
+  const { values: nextValues } = nextProps
+
+  if (!shallowEquals(nextValues, values)) {
+    return false
   }
 
-  shouldComponentUpdate(nextProps, nextState) {
-    const {values} = this.props;
-    const {values: nextValues} = nextProps;
+  // Since `values` has already been checked, we know they're not
+  // different, so the current `values` are carried over so the shallow
+  // equals comparison on the other props isn't affected by the `values`.
+  let nextPropsToCheck = {
+    ...nextProps,
+    values,
+  }
 
-    if (!shallowEquals(nextValues, values)) {
-      return true;
-    }
+  return shallowEquals(props, nextPropsToCheck)
+}
 
-    // Since `values` has already been checked, we know they're not
-    // different, so the current `values` are carried over so the shallow
-    // equals comparison on the other props isn't affected by the `values`.
-    let nextPropsToCheck = {
-      ...nextProps,
-      values,
-    };
+function FormattedMessage(props) {
+  const { formatMessage = defaultFormatMessage, textComponent: Text = 'span' } =
+    useIntl({ enforceContext: !props.defaultMessage }) || {}
+
+  const {
+    id,
+    description,
+    defaultMessage,
+    values,
+    children,
+    tagName: Component = Text,
+  } = props
+
+  let tokenDelimiter
+  let tokenizedValues
+  let elements
+
+  let hasValues = values && Object.keys(values).length > 0
+  if (hasValues) {
+    // Creates a token with a random UID that should not be guessable or
+    // conflict with other parts of the `message` string.
+    let uid = Math.floor(Math.random() * 0x10000000000).toString(16)
+
+    let generateToken = (() => {
+      let counter = 0
+      return () => `ELEMENT-${uid}-${(counter += 1)}`
+    })()
+
+    // Splitting with a delimiter to support IE8. When using a regex
```

Comment:
do we still support IE8?

Reply:
no, but i was trying to do a minimal port as first pass